Bluetooth: Host: Add encrypted advertising

theob-pro opened the pull request:
Create a new Bluetooth library and add Encrypted Advertising Data to it.

Encrypted Advertising Data is a new feature from the Bluetooth Core Specification 5.4. It provides a way to communicate encrypted data in advertising, scan response and EIR packets. To do that it introduce a new advertising data type called `Encrypted Advertising Data`. Also, it introduce a new characteristic called `Encrypted Data Key Material`, this provides a way to share the key material.

The library add two main functions `bt_ead_encrypt` and `bt_ead_decrypt`.

Also, add a new BabbleSim test that use the sample data from the Supplement to the Bluetooth Core Specification to validate the encryption, decryption and authentication.

Related issue: https://github.com/zephyrproject-rtos/zephyr/issues/55081

theob-pro committed:
Bluetooth: Host: Add Encrypted Advertising Data

Create a new Bluetooth library and add Encrypted Advertising Data to it.

Encrypted Advertising Data is a new feature from the Bluetooth Core
Specification 5.4. It provides a way to communicate encrypted data in
advertising, scan response and EIR packets. To do that it introduce a
new advertising data type called `Encrypted Advertising Data`. Also, it
introduce a new characteristic called `Encrypted Data Key Material`,
this provides a way to share the key material.

The library add two main functions `bt_ead_encrypt` and
`bt_ead_decrypt`.

Two helper functions are added to `bluetooth.h`. `bt_data_get_len` and
`bt_data_serialize`, the first one allow the user to get the total size
of a set of `bt_data` structures; the second one generate a spec
compliant bytes array from a `bt_data` structure. The last one is useful
because `bt_ead_encrypt` take as input those kind of bytes array.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Tests: Encrypted Advertising Data test

Add a new BabbleSim test that use the sample data from the Supplement to
the Bluetooth Core Specification to validate the Encrypted Advertising
Data feature implementation.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/gap.h:
```diff
@@ -68,6 +68,7 @@ extern "C" {
 #define BT_DATA_CSIS_RSI                0x2e /* CSIS Random Set ID type */
 #define BT_DATA_ADV_INT_LONG            0x2f /* Advertising Interval long */
 #define BT_DATA_BROADCAST_NAME          0x30 /* Broadcast Name */
+#define BT_DATA_ENCRYPTED_AD_DATA		0x31 /* Encrypted Advertising Data */
```
Use spaces instead of tabs for indention

Thalley reviewed (commented):
The second commit should be `tests: bluetooth:`

Why is this being added as a library, instead of just an API in `bluetooth.h`?

jori-nordic commented:
> Why is this being added as a library, instead of just an API in bluetooth.h?

The host is pretty bloated and tries to be too nice to the user. This pushes functionality that shouldn't be in the host to be inside the host (settings usage, a bunch of gatt functionality, adv resume comes to mind). That functionality then makes use of the internal state and functions, because why not, it's more efficient and that results in:
- more bugs, because more complex and internal state is used all over the place
- the API being not that nice to use in practice, because the user still has to be pretty familiar with the host's inner workings to use the 'user-friendly' functionality

That's why we'd like to slim down the host to a maintainable core + some libraries on top. 
Encrypted advertising can be implemented fully outside of the host, that's why we're making it a library.

hermabe commented:
The current tests only check that the advdata encrypts and decrypts back to itself. It would be nice to have some tests that checks that the encrypted data is correct as well. I suggest we use the "Encrypted Advertising Data Sample Data" from the CSS.

hermabe commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -3695,6 +3695,15 @@ struct bt_uuid_128 {
+ *  @brief GATT Characteristic Encrypted Data Key Material
+ */
+#define BT_UUID_GATT_ENC_DATA_KEY_MAT \
+	BT_UUID_DECLARE_16(BT_UUID_GATT_ENC_DATA_KEY_MAT_VAL)
```
These are already defined further down
```
/**
 *  @brief GATT Characteristic Encrypted Data Key Material UUID Value
 */
#define BT_UUID_GATT_EDKM_VAL 0x2b88
/**
 *  @brief GATT Characteristic Encrypted Data Key Material
 */
#define BT_UUID_GATT_EDKM \
	BT_UUID_DECLARE_16(BT_UUID_GATT_EDKM_VAL)
```

hermabe commented on lib/CMakeLists.txt:
```diff
@@ -9,3 +9,4 @@ add_subdirectory(os)
 add_subdirectory(util)
 add_subdirectory_ifdef(CONFIG_SMF smf)
 add_subdirectory_ifdef(CONFIG_OPENAMP              open-amp)
+add_subdirectory(bluetooth)
```
Does the library have to be in the `lib` folder or can it be moved to `subsys/bluetooth/host` so that it is in the same place as the rest of the bluetooth stack?

jhedberg replied:
`subsys/bluetooth/host` should be fine, I think. When this was originally described as a library (before I saw the pull request) I assumed that just meant that it would be a layer that builds on top of existing public Bluetooth APIs rather than using internal APIs, but I didn't expect to see something added to `lib/bluetooth/`

jori-nordic replied:
I'd like to keep it separated. Maybe `subsys/bluetooth/lib/` would be appropriate, we just thought that the root `lib/` directory existed for that exact purpose.

hermabe replied:
Not sure if there is any official guideline here, but the existing libraries in `lib/` seem to be general purpose. I would like to have all bluetooth related code in `subsys/bluetooth`.

jhedberg replied:
Yes, this will go under `subset/bluetooth/`. The main open question is whether it should be under `host` or some other directory there. I suppose adding a `lib` directory as suggested by @jori-nordic might make sense, but we should also look at what we already have and see if it makes sense to group this new functionality with existing Bluetooth sub-modules that can also be considered libraries (in the sense that there'd be multiple "library submodules" under the same directory.

hermabe commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,125 @@
+#include "bs_tracing.h"
+#include "bstests.h"
+#include "zephyr/toolchain/gcc.h"
+
```
All these common defines and data could be moved to a common header so that they dont have to be manually be kept in sync between the two files.

jhedberg commented on lib/bluetooth/encrypted_ad_data/encrypted_ad_data.h:
```diff
@@ -0,0 +1,76 @@
+#include <stddef.h>
+#include <stdint.h>
+
+#define EAD_AAD_SIZE	    1  /**< Additional Authenticated Data size in bytes */
+#define EAD_RANDOMIZER_SIZE 5  /**< Randomizer size in bytes */
+#define EAD_NONCE_SIZE	    13 /**< nonce size in bytes */
+#define EAD_KEY_SIZE	    16 /**< Key size in bytes */
+#define EAD_IV_SIZE	    8  /**< Initialisation Vector size in bytes */
+#define EAD_MIC_SIZE	    4  /**< MIC size in bytes */
+
```
If this is a public API it shouldn't be coming up with a new namespace like that (at least not without formally requesting/allocating one). The appropriate namespace should be `bt_*` like for all other Bluetooth APIs, and the header file should go in the same place as other public Bluetooth APIs.

jori-nordic commented:
> but we should also look at what we already have and see if it makes sense to group this new functionality with existing Bluetooth sub-modules that can also be considered libraries (in the sense that there'd be multiple "library submodules" under the same directory.

This sounds like it should be done in a separate PR. Having this `lib/` folder will make it easier to move functionality out of the 'core' host in the near future anyway.

Thalley commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+ * @return int Size of \p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
```
For constant-size arrays we've often used the size of the arrays in the declaration like
```suggestion
int bt_ead_encrypt_ad_data(const uint8_t session_key[16], const uint8_t iv[BT_EAD_IV_SIZE], const uint8_t *payload,
```

jhedberg replied:
Another point about the API name: both "ad" and "data" are already part of the "ead" acronym itself, so they're redundant. The appropriate name would therefore be simply `bt_ead_encrypt()`

Thalley commented on subsys/bluetooth/lib/encrypted_ad_data/encrypted_ad_data.c:
```diff
@@ -0,0 +1,108 @@
+	int encrypted_payload_size =
+		(encrypted_ad_data_size - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_AAD_SIZE;
+	int payload_size = encrypted_payload_size - BT_EAD_MIC_SIZE;
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/encrypted_ad_data/encrypted_ad_data.c:
```diff
@@ -0,0 +1,108 @@
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	const uint8_t *encrypted_payload =
+		encrypted_ad_data + BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE;
+	int encrypted_payload_size =
+		(encrypted_ad_data_size - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_AAD_SIZE;
+	int payload_size = encrypted_payload_size - BT_EAD_MIC_SIZE;
```
Hmm, this may be tricky if we want to return the payload_size as an `int` as well. 

I assume that we'll never see an overflow, but still. Perhaps `ssize_t`?

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,90 @@
+static uint8_t expected_data[] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};
+BUILD_ASSERT(sizeof(expected_data) == EXPECTED_DATA_SIZE);
+
+static uint8_t session_key[] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
+				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
+BUILD_ASSERT(sizeof(session_key) == BT_EAD_KEY_SIZE);
+static uint8_t iv[] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
+BUILD_ASSERT(sizeof(iv) == BT_EAD_IV_SIZE);
```
```suggestion
static uint8_t session_key[BT_EAD_KEY_SIZE] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
				
static uint8_t iv[BT_EAD_IV_SIZE] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,90 @@
+	if (!bt_addr_le_eq(addr, &prev_addr)) {
+		uint8_t *encrypted_data =
+			&ad->data[2]; /* the ad data are after the length and the type */
+		uint8_t encrypted_data_size = ad->len - 2; /* remove the type size */
```
Something like this is why I think that decrypting it in the stack may be more helpful; that way the users can treat encrypted and unencrypted data the same, and simply pass it to `bt_data_parse`, and don't have to worry about calculating the correct length to use here.

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,90 @@
+		uint8_t data_size =
+			((encrypted_data_size - BT_EAD_RANDOMIZER_SIZE) - BT_EAD_AAD_SIZE) -
+			BT_EAD_MIC_SIZE;
+		uint8_t *data = (uint8_t *)malloc(sizeof(uint8_t) * data_size);
```
Should we use `k_malloc` and `k_free` instead?

Thalley replied:
Not sure what the technical difference is, but it is suggested by the documentation: https://docs.zephyrproject.org/latest/develop/languages/c/index.html#dynamic-memory-management

aescolar replied:
Just as info: in the bsim target, by default malloc will use the underlaying linux malloc(). Meaning, you are allocating memory in the linux process heap. While k_malloc uses the Zephyr heap (which is actually static memory in the linux process).
If you want something that would run in the embedded target and/or emulate the embedded behaviour you want to use k_malloc/k_free. If you just want a buffer for bsim specific test purposes you can use whatever you prefer.

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,90 @@
+
+	LOG_DBG("Device found: %s (RSSI %d)", addr_str, rssi);
+
+	if (!bt_addr_le_eq(addr, &prev_addr)) {
+		uint8_t *encrypted_data =
+			&ad->data[2]; /* the ad data are after the length and the type */
```
How do we tell if the data is encrypted, before attempting to decrypt it?

theob-pro replied:
If we receive an advertising structure with the Ad Type `Encrypted Advertising Data`, then the Ad Data of this structure are encrypted.

Thalley replied:
Ah, nice. We should check the type before decrypting then, right?

Thalley replied:
What I suggest is that we take the `ad` here and forward it to `bt_data_parse`. Then in the `bt_data_parse` callback, we check the AD type to see if it is encrypted, and only then decrypt it. 

That would be more similar to how an application would/should use this, and would also avoid some of the length handling here.

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,90 @@
+
+void test_central(void)
+{
+	LOG_DBG("Central device");
+
+	bst_result = Passed;
+
+	int err;
```
```suggestion
	int err;
	
	LOG_DBG("Central device");

	bst_result = Passed;
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+#include "common.h"
+
+#define DATA_SIZE 6
+static uint8_t data[] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};
+BUILD_ASSERT(sizeof(data) == DATA_SIZE);
+
+static uint8_t session_key[] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
+				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
+BUILD_ASSERT(sizeof(session_key) == BT_EAD_KEY_SIZE);
+static uint8_t iv[] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
+BUILD_ASSERT(sizeof(iv) == BT_EAD_IV_SIZE);
```
```suggestion
static uint8_t data[DATA_SIZE] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};

static uint8_t session_key[BT_EAD_KEY_SIZE] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
				
static uint8_t iv[BT_EAD_IV_SIZE] = {0xFB, 0x56, 0xE1, 0xDA, 0xDC, 0x7E, 0xAD, 0xF5};
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	uint8_t *encrypted_ad_data = (uint8_t *)malloc(sizeof(uint8_t) * encrypted_data_size);
+
+	err = bt_ead_encrypt_ad_data(session_key, iv, data, DATA_SIZE, encrypted_ad_data);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	int err;
+	int encrypted_data_size =
+		BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE + DATA_SIZE + BT_EAD_MIC_SIZE;
+	uint8_t *encrypted_ad_data = (uint8_t *)malloc(sizeof(uint8_t) * encrypted_data_size);
```
Ditto with `k_malloc`/`k_free`

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	LOG_DBG("RPA expired");
+
+	int err;
+	int encrypted_data_size =
```
```suggestion
	uint8_t encrypted_data_size =
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	int err;
+	int32_t timeout = 0;
+	uint8_t num_events = 0;
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	start_params.num_events = num_events;
+
+	err = bt_le_ext_adv_start(adv, &start_params);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	params.interval_max = BT_GAP_ADV_SLOW_INT_MAX;
+
+	err = bt_le_ext_adv_create(&params, &adv_cb, adv);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+
+	if (err) {
+		FAIL("Failed to create advertiser (%d)", err);
+		return;
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+
+void test_peripheral(void)
+{
+	LOG_DBG("Peripheral device");
+
+	bst_result = Passed;
+
+	int err;
+	struct bt_le_ext_adv *adv = NULL;
```
```suggestion
	int err;
	struct bt_le_ext_adv *adv = NULL;
	
	LOG_DBG("Peripheral device");

	bst_result = Passed;
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	struct bt_le_ext_adv *adv = NULL;
+
+	err = bt_enable(NULL);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	create_adv(&adv);
+	start_adv(adv);
+
+	int encrypted_data_size =
+		BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE + DATA_SIZE + BT_EAD_MIC_SIZE;
+	uint8_t *encrypted_data = (uint8_t *)malloc(sizeof(uint8_t) + encrypted_data_size);
```
Declare at top of scope

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+	uint8_t *encrypted_data = (uint8_t *)malloc(sizeof(uint8_t) + encrypted_data_size);
+
+	err = bt_ead_encrypt_ad_data(session_key, iv, data, DATA_SIZE, encrypted_data);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+
+	if (err) {
+		FAIL("Failed to start advertiser (%d)", err);
+		return;
```
```suggestion
```

Alternatively is to add `return` after each `FAIL`

theob-pro commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+
+#include <zephyr/kernel.h>
+
+#define BT_EAD_AAD_SIZE	       1  /**< Additional Authenticated Data size in bytes */
+#define BT_EAD_RANDOMIZER_SIZE 5  /**< Randomizer size in bytes */
+#define BT_EAD_NONCE_SIZE      13 /**< nonce size in bytes */
+#define BT_EAD_KEY_SIZE	       16 /**< Key size in bytes */
+#define BT_EAD_IV_SIZE	       8  /**< Initialisation Vector size in bytes */
+#define BT_EAD_MIC_SIZE	       4  /**< MIC size in bytes */
```
What I have done is the way Doxygen do it. See [Doxygen Manual](https://www.doxygen.nl/manual/docblocks.html#memberdoc). I think it's better to stick to it.

Thalley replied:
I would strongly prefer to use the way that we do in Zephyr, than the suggestion in Doxygen :)

jhedberg replied:
Yes, documentation style should be consistent across all Zephyr APIs.

Btw, is all of the content of this heder file essential to be able to use the full public API? Anything that's not needed for public API usage would be better placed in the internal c-file or a (new) internal header file.

theob-pro commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
+BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
```
With your solution the array can be empty. With the build assert we ensure that the array is exactly the size we expect it to be.

Thalley replied:
What do you mean by `array can be empty.`? 

It can only be empty if `BT_EAD_AAD_SIZE`

alwa-nordic replied:
If the size of the array is specified in the declaration, it's not possible to check if there are fewer elements than expected. Any elements missing from the initializer will be static-initialized. For example:
```c
static uint8_t bt_ead_aad[BT_EAD_AAD_SIZE] = {};
/* Legal according to standard. No warnings. `bt_ead_aad[0]` is guaranteed to be `0`. */
/* Even, worse: The following check now passes.
 * For the check to work as expected, the size of the array must be implicitly
 * specified by the number of elements in the initializer, i.e. remove BT_EAD_AAD_SIZE
 * from the definition of bt_ead_aad. */
BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
```

Thalley replied:
> /* Even, worse: The following check now passes. */

Which is why it can be removed. 

The `BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);` doesn't verify correctness in the first place; it just verifies the size. If we want to ensure the size, then we can just use my suggestion which ensures size correctness (without the build assert), and while
```c
static uint8_t bt_ead_aad[BT_EAD_AAD_SIZE] = {};
```
would jus have `bt_ead_add[0] == 0`, it would give a build error if we attempted to do
```c
static uint8_t bt_ead_aad[BT_EAD_AAD_SIZE] = {0xEA, 0xEB};
```

My point was that the BUILD_ASSERT's purpose can be done in a simpler way (IMO)

Thalley replied:
Hmm, I see your point that in the case of the array being too small, it would be a difference, and we could forget to ensure the values. 

My suggestion can be ignored.

theob-pro commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+ * @note The function must be called each time the RPA is updated or the data
+ * are modified.
+ *
+ * @note The term `advertising strucure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
```
I use it in the documentation of the `payload`. I think it's useful as there is no explicit definition of such a structure (ie. Ad Len + Ad Type + Ad Data).

Thalley replied:
Ah, yes indeed; I was doing a search for it, but because of the typo in `structure` I didn't see it :)

Please fix the typo

jori-nordic commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+ * @return int Size of \p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+			   size_t payload_size, uint8_t *encrypted_ad_data);
```
That's a good point. But on the other hand, that's exactly why the host is so bloated. A lot of tiny features added to it because it 'makes sense' and we end up with a mess that's pretty hard to reason about and debug.

theob-pro replied:
Also, having it as an option as you describe cancel the option for the user to have both encrypted and non-encrypted ad data.

Thalley replied:
> Also, having it as an option as you describe cancel the option for the user to have both encrypted and non-encrypted ad data.

I am not fully aware of the details of the feature: Are you encrypting the advertising data, or just a single LTV structure in it?

Thalley replied:
> That's a good point. But on the other hand, that's exactly why the host is so bloated. A lot of tiny features added to it because it 'makes sense' and we end up with a mess that's pretty hard to reason about and debug.

I don't disagree, however from a user perspective it's much easier to use. It's basically a question of UX version how easy it is to maintain, right?

jori-nordic replied:
That's subjective I think. What you find easier to use and what I (or someone else) finds easier to use can be two very different things. Keep in mind that you and I have been working with this API for a while, and that it may cloud our judgment a bunch.

What I've seen is that in practice this nice UX promise ends up not being true: see the myriad of ways to call the GATT API - added for 'nicer UX' - ending up with footguns where the structs you pass are modified depending on what's in them, the stack behaving subtly differently when seemingly-unrelated kconfig options are enabled, stuff like that.

The advertising module is already complicated enough to use and maintain, as have shown some recent bugs.

My point of view, and I think @alwa-nordic and @theob-pro also agree, is that we should have a small and opinionated host API, that is used by pluggable host/lib modules. This way, the complexity and internal state re-use is minimized and the users can have a nice-to-use API with less bugs. 
The downside of course is the impact on code size and speed, but IMO it's better to have slower code that is correct rather than something that is fast but really buggy.

tldr: no, it's usually not that easier to use in practice and yes it makes it much harder to maintain.

theob-pro replied:
> I am not fully aware of the details of the feature: Are you encrypting the advertising data, or just a single LTV structure in it?

Only the Ad Data are encrypted, and placed in a new Ad Type `Encrypted Advertising Data`. The encrypted data can contain multiple advertising structure. The advertising payload can contain an encrypted advertising structure and a local name structure. The Ad Data of the encrypted advertising structure will be encrypted but not the one of the local name structure.

Thalley replied:
> tldr: no, it's usually not that easier to use in practice and yes it makes it much harder to maintain.

Indeed.

After @theob-pro 's helpful comment, I am fine with the current implementation. I had thought that the entire advertising data was encrypted, and not the individual LTV structures. So my suggestion is no longer valid, as that provides not only a less flexible solution, but also that it is easier to decode the data than I thought, as you simply just check the AD type after using `bt_data_parse` to decrypt the individual LTVs that may be encrypted. 

TLDR: I am fine with the suggested API

alwa-nordic commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+ * @return int Size of \p payload in bytes if the decryption is successful, else
+ *             return `-EIO` if an error occurred during the decryption.
+ */
+int bt_ead_decrypt_ad_data(const uint8_t *session_key, const uint8_t *iv,
+			   const uint8_t *encrypted_payload, size_t encrypted_ad_data_size,
+			   uint8_t *payload);
```
What you propose is dangerous. How will the application know if an AD is authenticated or not? A simple API is much easier to reason about for our users, and that benefits security. The intention is to have a companion library in the near future that handles the complexity that is in the test code in this PR. We want to define a simple API that can be built on top of. An API that is general because it's simple is better than an API that is general because it has complex extensions.

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/55225#discussion_r1119800205

> How will the application know if an AD is authenticated or not?

What is the use case for this? Why should encrypted advertising be considered differently than e.g. encrypted ACL data?

alwa-nordic replied:
> See [#55225 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/55225#discussion_r1119800205)
> 
> > How will the application know if an AD is authenticated or not?
> 
> What is the use case for this? Why should encrypted advertising be considered differently than e.g. encrypted ACL data?

Messages are authenticated to establish trust. Use case: An electronic shelf label must not accept updates that are unauthenticated, as that would open up for a trivial attack of just spoofing an advertisement with a plain-text unauthenticated update.

jhedberg commented:
@theob-pro I think it might make sense to rename the header file simply to `ead.h`, so that it's in line with the API naming. Probably same with the c-file. Application developers will be primarily referencing the generated documentation, so I don't think this will make the API any less discoverable for them.

jhedberg commented on subsys/bluetooth/lib/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: Apache-2.0
+
+add_subdirectory_ifdef(CONFIG_BT_ENCRYPTED_AD_DATA encrypted_ad_data)
```
Until the complexity of this new `subsys/bluetooth/lib/` directory increases, I'd just place the c-file directly there, rather than a subdirectory (and probably call it `ead.c` to be consistent with the API naming). We can always refactor this later into subdirectories if/when the complexity grows too large.

jori-nordic replied:
I don't agree. The point of my tirades in this PR is that the host is already complex enough, exactly because of this tendency that we have in Zephyr to force complexity downwards instead of having libraries/modules on top of simple subsystems. Seemingly simple features added over time with blurry boundaries have put us in this situation.

This PR strays from the way its done currently for good reason. When we have this lib/ folder, it will be easier to migrate stuff from the host into it.

jhedberg replied:
@jori-nordic did you perhaps misunderstand what I wrote? I'm 100% in favour of putting this in a new `subsys/bluetooth/lib/` directory, rather than `subsys/bluetooth/host/`. My point was that as long as the only content of `subsys/bluetooth/lib/ead/` is a single c-file, and the only content of `subset/bluetooth/lib` is a single `ead` subdirectory then I'd just put the c-file in `subsys/bluetooth/lib/ead.c` for now. Once we have more than, say 10 c-files, under lib, or if one library consists of multiple c-files, then we can group them under their own subdirectories as `subsys/bluetooth/lib/foo/`

jori-nordic replied:
Yep I did, sorry. I agree with what you said.

jhedberg commented on subsys/bluetooth/lib/encrypted_ad_data/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Nordic Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_ENCRYPTED_AD_DATA
```
It would probably make sense to make this follow the API naming as well, i.e. `config BT_EAD`.

jori-nordic replied:
but then it's harder to search for using the online kconfig browser or `menuconfig`

jhedberg replied:
Is it? Doesn't the search there work on the Kconfig symbol description and not the symbol name?

jori-nordic replied:
My bad it does. Searching for "Bluetooth" does give the `BT` symbol that has the description set to "Bluetooth".
I agree then.

alwa-nordic commented on include/zephyr/bluetooth/encrypted_ad_data.h:
```diff
@@ -0,0 +1,77 @@
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
```
Be careful about declaring variables in headers. Doing so requires header guards to prevent errors due to multiple definitions. And for sanity, any variables declared in header files must be static-const and should never have pointers to them. It's probably better to not declare variables in headers at all. (This will change when we start using C23 that has constexpr. :tada:)

theob-pro commented:
Rebased to fix the comments.

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ *             (1 byte).
+ * @param[out] payload Decrypted advertising structure. Can contain multiple
+	       advertising structure.
+ * @retval int Size of @p payload in bytes if the decryption is successful, else
+ *             return `-EIO` if an error occurred during the decryption.
```
```suggestion
 * @retval Size of @p payload in bytes if the decryption is successful.
 * @retval -EIO if an error occurred during the decryption.
```

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
+ * @retval int Size of @p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
```
```suggestion
 * @retval Size of @p encrypted_ad_data in bytes if successful
 * @retval -EIO if an error occurred during the encryption.
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,119 @@
+
+LOG_MODULE_REGISTER(bt_encrypted_ad_data, CONFIG_BT_EAD_LOG_LEVEL);
+
+static int bt_ead_generate_randomizer(uint8_t *randomizer)
```
```suggestion
static int bt_ead_generate_randomizer(uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE])
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,119 @@
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   int payload_size, uint8_t *encrypted_ad_data)
```
Please add argument validation (i.e. NULL checks, and `payload_size` probably can't be `0` or larger than 254).

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   int payload_size, uint8_t *encrypted_ad_data);
```
```suggestion
		   size_t payload_size, uint8_t *encrypted_ad_data);
```

theob-pro replied:
Why use `size_t`? The maximum size of the advertising data is 1650.

Thalley replied:
Two reasons:
1) It should be a unsigned type (we don't want negative payload sizes)
2) We use `size_t` everywhere else in the BT API to denote array/data lengths, except where it has a maximum size of 254 (e.g. `struct bt_data`)

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ *             return `-EIO` if an error occurred during the decryption.
+ */
+int bt_ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_payload,
+		   int encrypted_ad_data_size, uint8_t *payload);
```
```suggestion
		   size_t encrypted_ad_data_size, uint8_t *payload);
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,119 @@
+	return ead_size;
+}
+
+int bt_ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_ad_data,
+		   int encrypted_ad_data_size, uint8_t *payload)
```
Ditto with input validation checks

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/central.c:
```diff
@@ -0,0 +1,96 @@
+
+	LOG_DBG("Central device");
+
+	bst_result = Passed;
```
Aren't you setting the test to pass before even enabling the BT Stack here?

theob-pro replied:
Yes, this is intended. The test run for the whole simulation length. So by default, I set the test result to `Passed`. The test will fail only if it encounter a `FAIL(...)`.

Thalley replied:
Not sure I understand. 

Why not set the `bst_result` to `In_progress`, and when you get to the point where you want to pass the test, set it to `Passed`, like all the other BabbleSim tests we have? 

Why do you want to run it until the simulation length?

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+static uint8_t data[] = {0x5A, 0x65, 0x70, 0x68, 0x79, 0x72};
+BUILD_ASSERT(sizeof(data) == DATA_SIZE);
+
+static uint8_t session_key[] = {0xC0, 0xC1, 0xC2, 0xC3, 0xC4, 0xC5, 0xC6, 0xC7,
+				0xC8, 0xC9, 0xCA, 0xCB, 0xCC, 0xCD, 0xCE, 0xCF};
```
If the `session_key` needs to be identical between the central and peripheral, I suggest to put it (and any other common values) in a common file.

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/src/peripheral.c:
```diff
@@ -0,0 +1,137 @@
+
+	LOG_DBG("Peripheral device");
+
+	bst_result = Passed;
```
Ditto with setting `Passed` before even starting the test

theob-pro commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ * @retval int Size of @p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   int payload_size, uint8_t *encrypted_ad_data);
```
Actually we can't because `payload` can be multiple advertising structure concatenated together.
Using it for `encrypted_ad_data` could work but I in my opinion it's not the role of this function to create the advertising structure.

Thalley replied:
> because the data field of the bt_data structure is const *uint8_t. It mean that the user should still provide an adequate buffer.

Yes, you wouldn't be able to do in-line encryption, but that isn't possible anyways since you have `const uint8_t *payload`. Not sure why using `bt_data` is different here (for the input), but yes, we cannot use `bt_data` for the output.

theob-pro replied:
Yes, using `bt_data` for encryption is not a problem at all. But indeed, for the decryption, we can't output a `bt_data`. That create asymmetrical functions, I think that it can be confusing when you use the library.

Thalley replied:
That is a good point. Given the `constant` constraint, it may be impossible to use `bt_data` for this, if we want symmetrical functions. 

Since the output data is also LTV encoded, perhaps the output should be a `struct net_buf_simple *buf` so that it matches the data structure of the scanning `recv` callback, and makes it possible to forward to `bt_data_parse` to parse the decrypted data?

theob-pro replied:
> Since the output data is also LTV encoded, perhaps the output should be a `struct net_buf_simple *buf` so that it matches the data structure of the scanning `recv` callback, and makes it possible to forward to `bt_data_parse` to parse the decrypted data?

I tested that option but it's a bit irrelevant for `bt_ead_decrypt` to take a `net_buf_simple`. If you look at the test (https://github.com/zephyrproject-rtos/zephyr/pull/55225/files#diff-c8109c56879494f942802cad9eaa6352c80d4a4b5f61010de723adfa176665deR32), to decrypt the data I create a `net_buf_simple` but I only send the `data` field of the `net_buf_simple` to `bt_ead_decrypt`. And then I can easily send the `net_buf_simple` to `bt_data_parse`.

hermabe commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ * @note The term `advertising structure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
```
The session key is usually shared using a GATT characteristic, "Encrypted Data Key Material" from the core spec. But I guess you can share it however you want.

Thalley replied:
Alright, that makes sense. Is there anything that forbids devices to use the LTK?

theob-pro commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
+ * @param[in]  payload Advertising Data to encrypt. Can be multiple advertising
+ *             structure concatenated.
```
We don't know and it's not relevant for the function.

theob-pro commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the `nonce`.
+ * @param[in]  encrypted_payload Encrypted Advertising Data received. This
+ *             should only contain the advertising data and not the length nor
+ *             the type.
+ * @param[in]  encrypted_ad_data_size Size of the received advertising data in
+ *             bytes. Should be equal to the length minus the size of the type
+ *             (1 byte).
+ * @param[out] payload Decrypted advertising structure. Can contain multiple
+	       advertising structure.
```
The encrypted advertising feature is basically just a new advertising type. If you want to encrypt `ADS_1`, you will create a new advertising structure with the advertising structure of `ADS_1` (so, Ad Len + Ad Type + Ad Data) and you will encrypt that. 

Now if you want to encrypt multiple advertising structure, you just concatenate them. For example if you want to encrypt `ADS_1` and `ADS_3` you will create the following advertising structure:
```
ADS_4:
    Len: sizeof(ADS_1) + sizeof(ADS_3) + sizeof(ADS_4_Type) + sizeof(EAD_HDR) + sizeof(MIC)
    Type: 0x31 (Encrypted Advertising Data)
    Data: EAD_HDR + encrypt(ADS_1 + ADS_3 + MIC) // here + mean concatenate
```

So, when I write `This should only contain the advertising data and not the length nor the type.`, the length and the type are the one of `ADS_4`, the advertising structure with the encrypted data.

Thalley replied:
> So, when I write This should only contain the advertising data and not the length nor the type., the length and the type are the one of ADS_4, the advertising structure with the encrypted data.

Ah, that explains it :) So it is just the value of the LTV structure with the EAD AD type.

theob-pro commented:
Reworking the code of the lib and the test. Will ping concerned people when done.

theob-pro commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
```
Yes, the IV is part of the Key Material shared between the devices.

Thalley replied:
Yes, but is there any requirement for the user when generating this IV? If so, those requirement should be documented, and possible a helper function to generate IVs based on those requirements.

theob-pro replied:
No, there is no requirement on how to generate it. It's up to the upper layer to generate and update it as for the session key.

theob-pro commented:
> The current tests only check that the advdata encrypts and decrypts back to itself. It would be nice to have some tests that checks that the encrypted data is correct as well. I suggest we use the "Encrypted Advertising Data Sample Data" from the CSS.

Hello @hermabe, I updated the test to use the sample data from the CSS, could you have a look?

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+/**
+ * @brief Encrypt and authenticate the given advertising data. The resulting
+ * data in @p encrypted_payload will look like that:
+ * - Randomizer is added in the 5 first bytes;
```
Should we do something like 
```suggestion
 * - Randomizer is added in the @ref BT_EAD_RANDOMIZER_SIZE first bytes;
```
for these?

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,73 @@
+ * @param[in]  payload_size Size of the Advertising Data to encrypt.
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
```
@theob-pro this comment is still valid

theob-pro replied:
Sorry, I forgot to reply. I don't think that is necessary now. Also the "header" is only the Randomizer, the MIC is at the end. Anyway, now that there is the helpers functions it's no longer needed. Do you agree?

Thalley replied:
It was never "needed" :D But IMO it's easier to use a single value, rather than having to use two values to denote the additional size of the buffer. 

Alternatively we could also add a macro similar to `BT_ISO_SDU_BUF_SIZE` or `BT_L2CAP_BUF_SIZE` which are used for this exact same purpose.

theob-pro replied:
> But IMO it's easier to use a single value, rather than having to use two values to denote the additional size of the buffer.

I agree but with the `bt_ead_get_encrypted_data_size` and `bt_ead_get_decrypted_data_size` the user don't need to do calculation by itself. That said, I should maybe move out of the public header the variables that the user don't need anymore (`BT_EAD_AAD_SIZE`, `BT_EAD_RANDOMIZER_SIZE`, `BT_EAD_NONCE_SIZE`, `BT_EAD_MIC_SIZE`).

theob-pro replied:
Okay, I understand, I will add a macro to compute the EAD size based on the payload size.

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ * advertising data for a given set of `bt_data` structure.
+ *
+ * @param[in] ad_structs Array of `bt_data` structure.
+ * @param[in] num_elts Number of `bt_data` structure in @p ad_structs.
```
What is `elts` short for?

theob-pro replied:
Elements, I can write it completely, it will be more readable.

Thalley replied:
What about just `count`, `cnt, `length` or `len`, which are typically used for this purpose in the existing BT API?

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ * @brief Helper function to get the size (in bytes) of the encrypted
+ * advertising data for a given set of `bt_data` structure.
+ *
+ * @param[in] ad_structs Array of `bt_data` structure.
+ * @param[in] num_elts Number of `bt_data` structure in @p ad_structs.
```
Generally you can omit `in` and `out` if all arguments are `in`

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+		   size_t encrypted_payload_size, uint8_t *payload);
+
+/**
+ * @brief Helper function to get the size (in bytes) of the encrypted
```
```suggestion
 * @brief Get the size (in bytes) of the encrypted
```

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+size_t bt_ead_get_encrypted_data_size(const struct bt_data *ad_structs, size_t num_elts);
+
+/**
+ * @brief Helper function to get the size (in bytes) of the advertising
```
```suggestion
 * @brief Get the size (in bytes) of the advertising
```

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ * @brief Helper function to get the size (in bytes) of the advertising
+ * structure inside the AD Data of an encrypted advertising structure.
+ *
+ * @param[in] payload Advertising structure of type `BT_DATA_ENCRYPTED_AD_DATA`.
```
```suggestion
 * @param[in] payload Advertising structure of type @ref BT_DATA_ENCRYPTED_AD_DATA.
```

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+size_t bt_ead_get_decrypted_data_size(const struct bt_data *payload);
+
+/**
+ * @brief Helper function to convert `bt_data` structure into an advertising
+ * structure in an array. The data are formatted according to the Bluetooth Core
+ * Specification v. 5.4, vol. 3, part C, 11.
```
```suggestion
 * @brief Convert @ref bt_data structure into an advertising in an array. 
 *
 * The data are formatted according to the Bluetooth Core v. 5.4, vol. 3, part C, 11.
```

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ *
+ * @retval size_t Number of bytes written in @p output_ad_structure.
+ */
+size_t bt_data_to_ad_structure(const struct bt_data *input, uint8_t *output_ad_structure);
```
This seems like it should belong in `bluetooth.h` as it isn't related to encrypted advertising, but is more general. Suggest to define this close to the definition of `struct bt_data`. I assume that this can also be used internally in some of the placed where we convert `bt_data` to `uint8_t` buffer.

theob-pro replied:
I agree, a suggestion from @alwa-nordic was to also move `bt_ead_get_encrypted_data_size` and make it return only the size of the `bt_data` structures, then the user can use the macro `BT_EAD_SIZE` itself.

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ */
+size_t bt_data_to_ad_structure(const struct bt_data *input, uint8_t *output_ad_structure);
+
+#ifdef CONFIG_BT_TESTING
```
```suggestion
#if defined(CONFIG_BT_TESTING)
```

Not sure we should add test_specific code in the public APIs. It isn't uncommon for test application to access internal header files IIRC, so it could perhaps just be put there

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+ * @retval -EINVAL One of the argument is a NULL pointer.
+ * @retval -ECANCELED Error occurred during the random number generation.
+ */
+int bt_test_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *randomizer,
+			const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload);
```
I still suggest that for the fixed size arrays, we denote the size of the arrays in the function declaration

theob-pro replied:
Yes, I forgot to do it, but will do.

Thalley commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,147 @@
+int bt_test_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *randomizer,
+			const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload);
+
+#endif
```
```suggestion
#endif /* CONFIG_BT_TESTING */
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+	int err;
+
+	err = bt_rand(randomizer, BT_EAD_RANDOMIZER_SIZE);
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+
+static int bt_ead_generate_nonce(const uint8_t *iv, const uint8_t *randomizer, uint8_t *nonce)
+{
+	int err;
+	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];
+
+	if (randomizer == NULL) {
+		err = bt_ead_generate_randomizer(new_randomizer);
+
+		if (err != 0) {
```
```suggestion
	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];

	if (randomizer == NULL) {
		int err;
		
		err = bt_ead_generate_randomizer(new_randomizer);
		if (err != 0) {
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+
+size_t bt_ead_get_decrypted_data_size(const struct bt_data *payload)
+{
+	if (payload->type != BT_DATA_ENCRYPTED_AD_DATA) {
+		return 0;
+	}
+
+	size_t decrypted_data_size;
```
```suggestion
	size_t decrypted_data_size;
	
	if (payload->type != BT_DATA_ENCRYPTED_AD_DATA) {
		return 0;
	}

```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+	size_t ead_size = BT_EAD_RANDOMIZER_SIZE + payload_size + BT_EAD_MIC_SIZE;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+
+	err = bt_ccm_encrypt(session_key, nonce, payload, payload_size, bt_ead_aad, BT_EAD_AAD_SIZE,
+			     &encrypted_payload[BT_EAD_RANDOMIZER_SIZE], BT_EAD_MIC_SIZE);
+
+	LOG_HEXDUMP_DBG(encrypted_payload, ead_size, "Encrypted Data: ");
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
```
```suggestion
	if (err != 0) {
		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
		return -EIO;
	}
	
	LOG_HEXDUMP_DBG(encrypted_payload, ead_size, "Encrypted Data: ");
```

theob-pro replied:
IMO it can be useful to have the log, even if it's corrupted, before returning.

Thalley replied:
But if the function is returning an error, then the data isn't corrupted, it's (or may be) uninitialized, right?

theob-pro replied:
My bad, `bt_ccm_encrypt` return an error only for the arguments check so before writing anything on `encrypted_payload`. I will move the log.

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
```
Is it really valid to encrypt 0 octets of data?

theob-pro replied:
Nothing is prohibiting it. Not in the RFC of the CCM nor the CCM description in the Bluetooth Core.
I tried it and it will just create a Randomizer with a MIC. The MIC will be computed using the nonce and the AAD and so will be valid.

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+	const uint8_t *randomizer = encrypted_payload;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+
+	err = bt_ccm_decrypt(session_key, nonce, encrypted_ad_data, payload_size, bt_ead_aad,
+			     BT_EAD_AAD_SIZE, payload, BT_EAD_MIC_SIZE);
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+
+#endif
+
+int ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_payload,
```
```suggestion
static int ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_payload,
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+		return -EINVAL;
+	}
+
+	if (encrypted_payload_size < 0) {
+		LOG_ERR("encrypted_payload_size is set to %d. The minimum size of "
+			"encrypted_payload_size is BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE (%d).",
+			encrypted_payload_size, BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE);
+		return -EINVAL;
+	} else if (encrypted_payload_size == BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE) {
```
Can we just do `encrypted_payload_size < BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE`?

theob-pro replied:
Yes, you're right, I updated it. But yeah, I the second one is useful for the warning.

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,221 @@
+		return -EINVAL;
+	}
+
+	if (encrypted_payload_size < 0) {
```
`encrypted_payload_size` is unsigned, invalid check

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	int err;
+	int32_t timeout = 0;
+	uint8_t num_events = 0;
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+static void set_ad_data(struct bt_le_ext_adv *adv)
+{
+	int err;
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	start_params.num_events = num_events;
+
+	err = bt_le_ext_adv_start(adv, &start_params);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	params.interval_max = BT_GAP_ADV_SLOW_INT_MAX;
+
+	err = bt_le_ext_adv_create(&params, NULL, adv);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+
+	size_t size_ad_data = size_sample_ad_data;
+	size_t size_ead = BT_EAD_RANDOMIZER_SIZE + size_ad_data + BT_EAD_MIC_SIZE;
+
+	if (size_ead != size_sample_ead) {
+		LOG_ERR("Size of ead_1: %d\n", size_ead);
+		LOG_ERR("Size of sample_ead_1: %d", size_sample_ead);
+		FAIL("Computed size of encrypted data does not match the size of the encrypted "
+		     "data from the sample. (data set %d)\n",
+		     data_set);
+	}
+
+	uint8_t *ead = k_malloc(sizeof(ead[0]) * size_ead);
```
```suggestion
	uint8_t *ead;

	if (size_ead != size_sample_ead) {
		LOG_ERR("Size of ead_1: %d\n", size_ead);
		LOG_ERR("Size of sample_ead_1: %d", size_sample_ead);
		FAIL("Computed size of encrypted data does not match the size of the encrypted "
		     "data from the sample. (data set %d)\n",
		     data_set);
	}

	ead = k_malloc(sizeof(ead[0]) * size_ead);
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	uint8_t *ead = k_malloc(sizeof(ead[0]) * size_ead);
+
+	err = bt_test_ead_encrypt(session_key, iv, randomizer, sample_ad_data, size_ad_data, ead);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	if (err != 0) {
+		FAIL("Error during encryption.\n");
+	} else if (memcmp(ead, sample_ead, size_sample_ead) != 0) {
+		LOG_HEXDUMP_ERR(ead, size_ead, "Encrypted data from bt_ead_encrypt:");
+		LOG_HEXDUMP_ERR(sample_ead, size_sample_ead, "Encrypted data from sample:");
```
Does the `LOG_HEXDUMP_ERR` work? Have we registered a log?

Thalley replied:
Ah, I had missed the 
```c
LOG_MODULE_REGISTER(bt_bsim_ead_sample_data, CONFIG_BT_EAD_LOG_LEVEL);
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+		.data_len = size_ead, .type = BT_DATA_ENCRYPTED_AD_DATA, .data = ead};
+
+	err = bt_le_ext_adv_set_data(adv, &ead_struct, 1, NULL, 0);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+
+	LOG_HEXDUMP_DBG(ead, size_ead, "Encrypted data:");
+
+	struct bt_data ead_struct = {
```
Declare at top of scope

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,123 @@
+	LOG_DBG("Peripheral device. (data set %d)", data_set);
+
+	err = bt_enable(NULL);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/test_scripts/run_tests.sh:
```diff
@@ -0,0 +1,62 @@
+
+Execute ./bs_2G4_phy_v1 -v=${verbosity_level} -s="${simulation_id}_${data_set}" \
+    -D=2 -sim_length=60e6 $@
+
```
```suggestion

for process_id in $process_ids; do
    wait $process_id || let "exit_code=$?"
done

```

theob-pro replied:
Why do you want to do that?

Thalley replied:
You are running two tests here; without this they will be run in parallel, making it extremely hard to separate the logging from each test, thus making it hard to follow and to debug. 

IMO it is much better to run them separately

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/main.c:
```diff
@@ -0,0 +1,65 @@
+#include "bstests.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_bsim_ead_sample_data, CONFIG_BT_EAD_LOG_LEVEL);
```
I'm wondering if this should use this same log level, as the library. 

Surely they are related, but it may be the case that we want to log debug information from the test, without logging debug information from the stack. 

For now it's not a problem, but just something I was wondering :)

theob-pro commented:
In the latest rebase:

- moved `bt_data_to_ad_structure` from `ead.h` to `bluetooth.h`;
- moved `bt_ead_get_encrypted_data_size` from `ead.h` to `bluetooth.h` and renamed it `bt_get_bt_data_size`;
- removed `bt_ead_get_decrypted_data_size`;
- added `BT_DECRYPTED_PAYLOAD_SIZE` macro to replace `bt_ead_get_decrypted_data_size`;
- moved `BT_EAD_NONCE_SIZE` and `BT_EAD_AAD_SIZE` from `ead.h` to `ead.c`;
- updated the test according to the previous modification.

theob-pro commented:
Rebased to update commit message.

jori-nordic commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+
+		struct net_buf_simple *decrypted_buf = k_malloc(sizeof(struct net_buf_simple));
+
+		decrypted_buf->size = decrypted_data_size;
```
don't think you have to use a runtime alloc. Since you already know the size (CSS data) then you can create a static one with `NET_BUF_SIMPLE_DEFINE_STATIC(_name, _size)`. If you want to keep the runtime alloc, I would advise to at least move it to its own fn like so:
```c
static struct net_buf_simple* alloc_buf(size_t size)
{
	struct net_buf_simple *buf = k_malloc(sizeof(struct net_buf_simple));

	buf->size = size;
	buf->__buf = k_malloc(sizeof(buf->data[0]) * size);
	net_buf_simple_reset(buf);

	return buf;
}

static void free_buf(struct net_buf_simple *buf)
{
	k_free(buf->__buf);
	k_free(buf);
}
```

Thalley replied:
Agreed. That way you can also remove the heap if you want.

jori-nordic commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+
+		if (err != 0) {
+			FAIL("Error during decryption.\n");
+		} else if (memcmp(decrypted_buf->data, sample_ad_data_1, decrypted_data_size)) {
```
why isn't the test failing?

theob-pro commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+		return -EIO;
+	}
+
+	LOG_HEXDUMP_DBG(encrypted_ad_data, encrypted_ad_data_size, "Encrypted Data: ");
```
You're right for decryption, it can be useful to log the decrypted data before returning. For encryption it's not useful, as discussed with @Thalley previously the `bt_ccm_encrypt` return only before doing any encryption, at the args check.

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -389,6 +389,34 @@ struct bt_data {
+ * set of @ref bt_data structure.
+ *
+ * @param[in] ad_structs Array of @ref bt_data structure.
+ * @param[in] ad_structs_len Number of @ref bt_data structure in @p ad_structs.
```
In this sentence, the entire word can just be remove and it still works IMO

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	if (encrypted_payload_size < BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE) {
+		LOG_ERR("encrypted_payload_size is set to %d. The minimum size of "
```
Could also make this a LOG_DBG. If the feature and the API allows this, it's not really something we should warn about

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+int bt_ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		   const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
```
I'd suggest something like this:
```suggestion
	CHECKIF(session_key == NULL) { 
		LOG_DBG("session_key is NULL");
		return -EINVAL;
	}

	CHECKIF(iv == NULL) {
		LOG_DBG("iv is NULL);
		return -EINVAL;
	}

	CHECKIF(payload == NULL) {
		LOG_DBG("payload is NULL);
		return -EINVAL;
	}

	CHECKIF(encrypted_payload == NULL) {
		LOG_DBG("encrypted_payload is NULL);
		return -EINVAL;
	}
```

Since we are returning an error, we shouldn't use LOG_ERR. The splitting of the validation checks make it much easier to, when LOG_DBG is enabled, to see exactly which one of them is failing. Finally, using CHECKIF allows the user to disable this check or make it an ASSERT if wanted.

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+			const uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE], const uint8_t *payload,
+			size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
```
Ditto here w.r.t using `CHECKIF` and `LOG_DBG`

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
```
We should consider using `LOG_DBG` for these instead.

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+	const uint8_t *encrypted_ad_data = &encrypted_payload[BT_EAD_RANDOMIZER_SIZE];
+	size_t encrypted_ad_data_size = encrypted_payload_size - BT_EAD_RANDOMIZER_SIZE;
+	size_t payload_size = encrypted_ad_data_size - BT_EAD_MIC_SIZE;
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,220 @@
+		   const uint8_t *encrypted_payload, size_t encrypted_payload_size,
+		   uint8_t *payload)
+{
+	if (session_key == NULL || iv == NULL || encrypted_payload == NULL || payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
```
Ditto

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+			     data_set);
+		}
+
+		struct net_buf_simple *decrypted_buf = k_malloc(sizeof(struct net_buf_simple));
```
Declare at top of scope

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+	} else {
+		LOG_DBG("Parsed data:");
+		LOG_DBG("len : %d", data->data_len);
+		LOG_DBG("type: %s", bt_hex(&data->type, 1));
```
```suggestion
		LOG_DBG("type: 0x%02x", data->type);
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+	int err;
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, device_found);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+	LOG_DBG("Central device. (data set %d)", data_set);
+
+	err = bt_enable(NULL);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,112 @@
+		k_free(decrypted_buf->__buf);
+		k_free(decrypted_buf);
+
+		PASS("Central test passed. (data set %d)\n", data_set);
```
Since we are passing this test in this case, we could return `false` to stop parsing any more AD structures

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/peripheral.c:
```diff
@@ -0,0 +1,126 @@
+	size_t size_ead = BT_EAD_ENCRYPTED_PAYLOAD_SIZE(size_ad_data);
+
+	if (size_ead != size_sample_ead) {
+		LOG_ERR("Size of ead_1: %d\n", size_ead);
+		LOG_ERR("Size of sample_ead_1: %d", size_sample_ead);
```
```suggestion
		LOG_ERR("Size of ead_1: %zu\n", size_ead);
		LOG_ERR("Size of sample_ead_1: %zu", size_sample_ead);
```

theob-pro commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,126 @@
+ * @retval 0 Data have been correctly encrypted and authenticated.
+ * @retval -EIO Error occurred during the encryption or the authentication.
+ * @retval -EINVAL One of the argument is a NULL pointer.
+ * @retval -ECANCELED Error occurred during the random number generation.
```
No, I will remove the `retval`.

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,115 @@
+			FAIL("Received Randomizer does not match the expected one.\n");
+		}
+
+		decrypted_buf.size = decrypted_data_size;
+		decrypted_buf.__buf = decrypted_payload;
```
```suggestion
		net_buf_simple_init_with_data(&decrypted_buf, decrypted_payload, decrypted_data_size);
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/central.c:
```diff
@@ -0,0 +1,115 @@
+
+		err = bt_ead_decrypt(sample_data->session_key, sample_data->iv, data->data,
+				     data->data_len, decrypted_buf.data);
+
```
```suggestion
```

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/common.c:
```diff
@@ -0,0 +1,36 @@
+		data_set = 1;
+	}
+
+	if (data_set == 1) {
+		sample_data = &sample_data_1;
+	} else if (data_set == 2) {
+		sample_data = &sample_data_2;
+	}
```
Have you considered putting the sample_data in an array, and using the argument as an index?

Thalley commented on tests/bluetooth/bsim/host/adv/encrypted/css_sample_data/src/common.c:
```diff
@@ -0,0 +1,36 @@
+		data_set = 1;
+	}
+
+	if (data_set == 1) {
+		sample_data = &sample_data_1;
+	} else if (data_set == 2) {
+		sample_data = &sample_data_2;
+	}
```
What happens if someone sets `data_set = 3`?

jori-nordic replied:
why would someone do that? :p
I think more bsim tests (and tests in general) are not coded to be up to the same standards as applications, as they are not intended to be used by end-users.

anyway you have your answer a few lines above in this case.

theob-pro replied:
Indeed, if `data_set` is different from 1 or 2, it will default to 1.

Thalley reviewed (changes requested):
A few comments that should be addressed, but mostly nitpicks and formatting comments otherwise. 

Implementation and API-wise I think it's looking pretty good, and since it's experimentally marked, it should be fine. 

One other note @theob-pro as per the https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html please don't resolve other peoples comments. The author of the comment should do that.

theob-pro commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -389,6 +389,34 @@ struct bt_data {
+ *             structure in @p input. The size of it must be at least the size
+ *             of the `input->data_len + 2` (for the type and the length).
+ *
+ * @retval size_t Number of bytes written in @p output_ad_structure.
```
The first argument of `@retval` is the return value, the second is the description.

Doxygen reference: https://www.doxygen.nl/manual/commands.html#cmdretval

Thalley replied:
Right, but `size_t` isn't a value; that's a type. 

In this case we probably should just use `@return`

theob-pro commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -389,6 +389,34 @@ struct bt_data {
+ *
+ * @retval size_t Number of bytes written in @p output_ad_structure.
+ */
+size_t bt_serialize_bt_data(const struct bt_data *input, uint8_t *output_ad_structure);
```
I don't think it's necessary, what kind of error would you like to handle?

Thalley replied:
If any of the values are NULL for example :) But may be overkill for this function

theob-pro replied:
Yes it's not necessary here. If one of the input is `NULL`, the returned value will be 0. According to the documentation the result is the number of bytes written in `output_ad_structure`. That make sense then to have a 0 value.

jori-nordic commented on include/zephyr/bluetooth/ead.h:
```diff
@@ -0,0 +1,127 @@
+		   const uint8_t *encrypted_payload, size_t encrypted_payload_size,
+		   uint8_t *payload);
+
+#if defined(CONFIG_BT_TESTING)
```
I think it would be acceptable to have an internal header file, yes.

alwa-nordic commented on subsys/bluetooth/lib/ead.c:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will only contain the "
+			"Randomizer and the MIC.");
```
I agree. It's possible there may be use-cases for an empty EAD. We shouldn't impose artificial restrictions.

Thalley replied:
@theob-pro will you change the WRN to DBG?

theob-pro replied:
No, sorry I tough that our separate discussion resolved this issue. I agreed with you that for recoverable error, `LOG_ERR` was not suitable but here, `LOG_WRN` exactly fit the Zephyr documentation.

theob-pro commented:
Rebased to resolve file conflict (`tests/bluetooth/bsim/host/compile.sh`).

theob-pro commented:
Rebased to resolve file conflict (`tests/bluetooth/bsim/host/compile.sh`).

Thalley reviewed (dismissed):
My remaining comments are about undocumented/required formatting changes, and one comment about a LOG_WRN that I think should be LOG_DBG, but it's not bad enough that I will block this PR anymore :)

jori-nordic commented:
I agree, let's get this in, and @theob-pro and @Thalley can discuss the small stuff in a new PR.

jhedberg reviewed (changes requested):
Sorry for being slow to review this. I've tried to keep my comments restricted to things which require more effort/pain to fix later in the game, i.e. the API details.

jhedberg commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -388,6 +388,34 @@ struct bt_data {
+ *
+ * @return Number of bytes written in @p output_structure.
+ */
+size_t bt_data_serialize(const struct bt_data *input, uint8_t *output_structure);
```
minor, but could you just call it `output`? You seem to have a fondness for long symbol names, but "structure" doesn't really add any value here. In fact, it's slightly confusing, since while `input` is a struct, `output_structure` is not :)

jori-nordic replied:
'structure' is what the BT spec calls the serialized AD data
